zerver/views/tusd.py

```diff
@@ -0,0 +1,100 @@
+import os
+import random
+from shutil import move
+from typing import Any, Dict, Mapping
+
+from django.conf import settings
+from django.http import HttpRequest, HttpResponse
+from django.utils.translation import gettext as _
+
+from zerver.decorator import internal_notify_view
+from zerver.lib.exceptions import JsonableError
+from zerver.lib.request import REQ, has_request_variables
+from zerver.lib.response import json_success
+from zerver.lib.upload import check_upload_within_quota
+from zerver.lib.upload.base import create_attachment, sanitize_name
+from zerver.lib.upload.s3 import S3UploadBackend
+from zerver.models import UserProfile
+
+
+def generate_path(realm_id: str, file_id: str, uploaded_file_name: str) -> str:
+    return "/".join(
+        [
+            realm_id,
+            format(random.randint(0, 255), "x"),
+            file_id,
+            sanitize_name(uploaded_file_name),
+        ]
+    )
+
+
+def move_s3_file(path_id: str, key: str) -> bool:
+    upload_backend = S3UploadBackend()
+    return upload_backend.move_file_inside_bucket(key, path_id)
+
+
+def move_local_file(current_path: str, path_id: str) -> None:
+    assert settings.LOCAL_FILES_DIR is not None
+    file_path = os.path.join(settings.LOCAL_FILES_DIR, path_id)
+
+    os.makedirs(os.path.dirname(file_path), exist_ok=True)
+
+    move(current_path, file_path)
+
+
+def handle_pre_create_hook(
+    request: HttpRequest, user_profile: UserProfile, data: Dict[str, Any]
+) -> HttpResponse:
+    file_size = data["Size"]
+    assert file_size != 0
+    if settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 < file_size:
+        raise JsonableError(
+            _("Uploaded file is larger than the allowed limit of {} MiB").format(
+                settings.MAX_FILE_UPLOAD_SIZE,
+            )
+        )
+    check_upload_within_quota(user_profile.realm, file_size)
```

Comment:
We are creating attachment in pre-finish hook, let's assume user is uploading 1GB file and failed to upload at last patch request. The storage isn't in DB so that isn't counting here.

Here is the another edge case for here, assume remaining upload quota is 1GB, and 2 users start 1GB file upload at same time, than it will allow both of them. 
Maybe we can call this function in `pre-finish` hook also.

Reply:
https://github.com/tus/tusd/issues/435 is where they are discussing about expiration extension. It might not roll out in v2 so we have to do our own implementation. We can run a cron job which will remove the files created let's say 1 day. There will be very low chances that someone will try to upload the file which failed a day before.

Reply:
Yeah I think we don't need a new cron job -- likely just a new function called from `delete_old_unclaimed_attachments`, with tests, for doing garbage-collection of those older than some setting, say `settings.TUSD_MAX_UPLOAD_TIME`, which we can default to 1 day (but should surely be a setting defined in `default_settings.py`)